animations-dbus/animations-dbus-server-surface.c

```diff
@@ -0,0 +1,691 @@
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static GVariant *
+serialize_attached_effects_to_variant (GHashTable *effects_for_events)
+{
+  g_auto(GVariantDict) vardict;
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_variant_dict_init (&vardict, NULL);
+
+  g_hash_table_iter_init (&iter, effects_for_events);
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      const char *event = key;
+      GQueue *effects = value;
+      g_autoptr(GPtrArray) names = g_ptr_array_new_full (g_queue_get_length (effects), NULL);
+
+      for (GList *link = g_queue_peek_head_link (effects);
+           link != NULL;
+           link = link->next)
+        {
+          AttachedEffectInfo *info = link->data;
+          g_ptr_array_add (names,
+                           (gpointer) g_dbus_interface_skeleton_get_object_path (G_DBUS_INTERFACE_SKELETON (info->server_effect)));
+        }
+
+      g_variant_dict_insert_value (&vardict,
+                                   event,
+                                   g_variant_new_strv ((const char * const *) names->pdata,
+                                                       names->len));
+    }
+
+  return g_variant_dict_end (&vardict);
+}
+
+static void
+animations_dbus_server_surface_get_property (GObject    *object,
+                                             guint       prop_id,
+                                             GValue     *value,
+                                             GParamSpec *pspec)
+{
+  AnimationsDbusServerSurface *server_surface = ANIMATIONS_DBUS_SERVER_SURFACE (object);
+  AnimationsDbusServerSurfacePrivate *priv =
+    animations_dbus_server_surface_get_instance_private (server_surface);
+
+  switch (prop_id)
+    {
+    case PROP_CONNECTION:
+      g_value_set_object (value, priv->connection);
+      break;
+    case PROP_SERVER:
+      g_value_set_object (value, priv->server);
+      break;
+    case PROP_BRIDGE:
+      g_value_set_object (value, priv->bridge);
+      break;
+    case PROP_TITLE:
+      g_value_set_string (value,
+                          animations_dbus_server_surface_bridge_get_title (priv->bridge));
+      break;
+    case PROP_GEOMETRY:
+      g_value_set_variant (value,
+                           animations_dbus_server_surface_bridge_get_geometry (priv->bridge));
+      break;
+    case PROP_EFFECTS:
+      g_value_set_variant (value,
+                           serialize_attached_effects_to_variant (priv->attached_effects_for_events));
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static void
+animations_dbus_server_surface_dispose (GObject *object)
+{
+  AnimationsDbusServerSurface *server_surface = ANIMATIONS_DBUS_SERVER_SURFACE (object);
+  AnimationsDbusServerSurfacePrivate *priv = animations_dbus_server_surface_get_instance_private (server_surface);
+
+  priv->server = NULL;
```

Comment:
It isn't strictly speaking necessary - I just wanted to make it clear that `server` is a non-owning reference (perhaps it may be a good usecase for `GWeakRef`).

Reply:
No need to use `GWeakRef` unless you need the notification callback about the weak ref being cleared to `NULL` when the object is finalised.

Personally, I think it would be marginally clearer just to have a comment in this `dispose` function:
```
/* Don’t clear priv->server as it’s not owned */
```

Reply:
Fair enough - I can just have a comment./